crates/data_components/src/databricks/spark_connect.rs

```diff
@@ -17,30 +17,60 @@ limitations under the License.
 use arrow::datatypes::SchemaRef;
 use async_trait::async_trait;
 use datafusion::{datasource::TableProvider, sql::TableReference};
-use std::{error::Error, sync::Arc};
+use std::sync::Arc;
 use uuid::Uuid;
 
+use crate::token_provider::TokenProvider;
 use crate::{Read, spark_connect::SparkConnect};
 
 #[derive(Clone)]
 pub struct DatabricksSparkConnect {
-    spark_connect: SparkConnect,
+    spark_connect: Arc<SparkConnect>,
 }
+
 impl DatabricksSparkConnect {
     pub async fn new(
         endpoint: String,
         cluster_id: String,
         token: String,
         databricks_use_ssl: bool,
-    ) -> Result<Self, Box<dyn Error + Send + Sync>> {
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
         let session_id = Uuid::new_v4();
         let connection = format!(
             "sc://{endpoint}:443/;use_ssl={databricks_use_ssl};user_id=spice.ai;session_id={session_id};token={token};x-databricks-cluster-id={cluster_id}"
         );
         Ok(Self {
-            spark_connect: SparkConnect::from_connection(connection.as_str()).await?,
+            spark_connect: Arc::new(SparkConnect::from_connection(connection.as_str()).await?),
         })
     }
+
+    pub async fn new_m2m(
+        endpoint: String,
+        cluster_id: String,
+        databricks_use_ssl: bool,
+        token_provider: Arc<dyn TokenProvider>,
+    ) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
+        let token = token_provider.get_token().await?;
```

Comment:
Could reduce code duplication of L55-60 via:

```rust
pub async fn new_m2m(
    endpoint: String,
    cluster_id: String,
    databricks_use_ssl: bool,
    token_provider: Arc<dyn TokenProvider>,
) -> Result<Self, Box<dyn std::error::Error + Send + Sync>> {
  let token = token_provider.get_token().await?;
  let result = self.new(endpoint, cluster_id, token, databricks_use_ssl)?;
...
```

Reply:
Thanks!